hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/impl/SqlClientPool.java

```diff
@@ -355,15 +399,22 @@ public CompletionStage<Void> rollbackTransaction() {
 			return connection().thenCompose( ReactiveConnection::rollbackTransaction );
 		}
 
-		@Override
 		public ReactiveConnection withBatchSize(int batchSize) {
-			if ( connection == null ) {
-				this.batchSize = batchSize;
+			if ( closed ) {
+				throw LOG.connectionIsClosed();
+			}
+
+			if ( connectionFuture.isDone() ) {
+				// connection exists, we can let callers use the delegate and forget about the proxy.
+				return connectionFuture.getNow( null ).withBatchSize( batchSize );
```

Comment:
Note the pattern `if ( ... .isDone() ) { get() }` should be avoided in general, because it's sensitive to [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use). That's why I suggested `var value = getNow(null); if ( value != null)` .

Though in this case... I think the worst that could happen is a missed opportunity, so it's probably fine.